core/src/main/scala/kafka/coordinator/transaction/ProducerIdManager.scala

```diff
@@ -75,87 +61,179 @@ trait ProducerIdGenerator {
           }
 
-          ProducerIdBlock(brokerId, currProducerIdBlock.blockEndId + 1L, currProducerIdBlock.blockEndId + ProducerIdManager.PidBlockSize)
+          new ProducerIdsBlock(brokerId, currProducerIdBlock.producerIdEnd + 1L, ProducerIdsBlock.PRODUCER_ID_BLOCK_SIZE)
         case None =>
-          debug(s"There is no producerId block yet (Zk path version $zkVersion), creating the first block")
-          ProducerIdBlock(brokerId, 0L, ProducerIdManager.PidBlockSize - 1)
+          logger.debug(s"There is no producerId block yet (Zk path version $zkVersion), creating the first block")
+          new ProducerIdsBlock(brokerId, 0L, ProducerIdsBlock.PRODUCER_ID_BLOCK_SIZE)
       }
 
-      val newProducerIdBlockData = ProducerIdManager.generateProducerIdBlockJson(currentProducerIdBlock)
+      val newProducerIdBlockData = ProducerIdBlockZNode.generateProducerIdBlockJson(newProducerIdBlock)
 
       // try to write the new producerId block into zookeeper
-      val (succeeded, version) = zkClient.conditionalUpdatePath(ProducerIdBlockZNode.path,
-        newProducerIdBlockData, zkVersion, Some(checkProducerIdBlockZkData))
+      val (succeeded, version) = zkClient.conditionalUpdatePath(ProducerIdBlockZNode.path, newProducerIdBlockData, zkVersion, None)
       zkWriteComplete = succeeded
 
-      if (zkWriteComplete)
-        info(s"Acquired new producerId block $currentProducerIdBlock by writing to Zk with path version $version")
+      if (zkWriteComplete) {
+        logger.info(s"Acquired new producerId block $newProducerIdBlock by writing to Zk with path version $version")
+        return newProducerIdBlock
+      }
     }
+    throw new IllegalStateException()
   }
+}
 
-  private def checkProducerIdBlockZkData(zkClient: KafkaZkClient, path: String, expectedData: Array[Byte]): (Boolean, Int) = {
-    try {
-      val expectedPidBlock = ProducerIdManager.parseProducerIdBlockData(expectedData)
-      zkClient.getDataAndVersion(ProducerIdBlockZNode.path) match {
-        case (Some(data), zkVersion) =>
-          val currProducerIdBLock = ProducerIdManager.parseProducerIdBlockData(data)
-          (currProducerIdBLock == expectedPidBlock, zkVersion)
-        case (None, _) => (false, -1)
-      }
-    } catch {
-      case e: Exception =>
-        warn(s"Error while checking for producerId block Zk data on path $path: expected data " +
-          s"${new String(expectedData, StandardCharsets.UTF_8)}", e)
-        (false, -1)
-    }
+class ZkProducerIdManager(brokerId: Int,
+                          zkClient: KafkaZkClient) extends ProducerIdGenerator with Logging {
+
+  private var currentProducerIdBlock: ProducerIdsBlock = ProducerIdsBlock.EMPTY
+  private var nextProducerId: Long = -1L
+
+  // grab the first block of producerIds
+  this synchronized {
+    getNewProducerIdBlock()
+    nextProducerId = currentProducerIdBlock.producerIdStart
+  }
+
+  private def getNewProducerIdBlock(): Unit = {
```

Comment:
It's probably better to make this function synchronized too, to emphasize that `currentProducerIdBlock` must be modified under the lock.

Also, why does the initial synchronized block set `nextProducerId` to `currentProducerIdBlock.producerIdStart`, while the `generateProducerId` function sets it to `currentProducerIdBlock.producerIdStart + 1`

Reply:
I don't actually need to set it to `currentProducerIdBlock.producerIdStart` since it's already initialized to -1 above. I'll remove that. 

I'll add the `synchronized` here as well